In [1]:
import pandas as pd
import numpy as np
from collections import Counter

#### Теория

Пусть $X = (x_1, ..., x_n)$ выборка из неизвестного распределения $ P \in (P_\theta | \theta \in \Theta) $, где \

1. либо все $ P_\theta $ абсолютно непрерывны и $ p_\theta(x) $ - плотность $ P_\theta $ \
2. либо все $ P_\theta $ дискретны и $ p_\theta(x) = P_\theta(X_1 = x) вероятности$, \

тогда, \

$ L_X(\theta) = p_\theta(x) = \prod\limits_{i = 1}^n p_\theta(x_i) $ - функция правдоподобия \

$l_X(\theta) = ln L_X(\theta)$ - логарифмическая функция правдоподобия \

Смысл: вероятность выборки в зависимости от значения параметра, степень доверия к конкретному значению параметра $\theta$

Оценка максимального правдоподобия:

$\hat{\theta} = \underset{\theta \in \Theta}{\operatorname{argmax}} L_X(\theta) $

Свойства (если функция правдоподобия непрерывно дифференцируема): \
1. сильно состоятельна
2. ассимптотически эффективна
3. ассимптотически нормальна

#### Задача №1. Распределение Пуассона

По выборке (2,  1,  0,  0,  1) из генеральной совокупности, имеющей  закон распределения Пуассона, найдите значение логарифмической функции правдоподобия для $\lambda=3$.

**Решение:**

Найдем логарифмическую функцию правдоподобия Пуассоновского распределения. Т.к. пуассоновское распределение дискретно, тогда $ p_\theta = P_\theta(X_1 = x) $ в случае Пуассоновского распределения:

$ p_\lambda(x) = P(X_1 = x) = \frac{\lambda^x}{x!}exp^{-\lambda} $,

тогда функция правдоподобия будет: 

$L_X(\lambda) = \prod\limits_{i = 1}^n p_\lambda(x) = \prod\limits_{i = 1}^n \frac{\lambda^x}{x!}exp^{-\lambda} = \frac{ \lambda^{\sum\limits_{i=1}^n x_i} }{\prod\limits_{i = 1}^n x_i!}  exp^{-n\lambda}$ 

а логарифмическая функция правдоподобия: 

$l_X(\lambda) = ln(\frac{ \lambda^{\sum\limits_{i=1}^n x_i} }{\prod\limits_{i = 1}^n x_i!}  exp^{-n\lambda}) = \sum\limits_{i=1}^n x_i \cdot ln \lambda - \sum\limits_{i=1}^n ln(x_i)! - n\lambda$ \

подставим наши данные

$l_X(3) = 2+1+0+0+1 \cdot ln(3) - (ln(2!)+ln(1!)+ln(0!)+ln(0!)+ln(1!)) - 5 \cdot 3 = 4,4 - 0,69 - 15 = -11,29$

**Ответ:**

$l_X(3) = -11,29$

#### Задача №2. Равномерное распределение

По предложенной выборке (-12.8, -8.85, -13.45, -5.17, -12.11, -16.13, -7.57, -1.64, -8.91, -6.62) с помощью метода максимального правдоподобия найдите точечную оценку параметра $\theta$ равномерно на $[-\theta, 0]$ распределенной случайной величины 

$ f_\xi(x) = \begin{equation*}
              \begin{cases}
                \frac{1}{\theta} , x \in [-\theta, 0],
                \\
                0 , x \notin [-\theta, 0].
              \end{cases}
              \end{equation*}$



**Решение:**

Т.к. мы имеем равномерное непрерывное распределение, то плотность распределения будет

$p_\theta = \frac{1}{b-a}$ в нашем случае $a = - \theta$, а $b = 0$

тогда, функция правдоподобия будет

$L_X(\theta) = \prod\limits_{i = 1}^n \frac{1}{0-(-\theta)} = \frac{1}{ \theta^n} $

Чтобы найти оценку функции правдоподобия, нам нужно найти argmax, аргумент при котором функция достигает максимума. т.к. по условию задачи, оценка должна принадлежать интервалу $[-\theta, 0]$, то функция правдоподобия достигает максимума в аргументе $x_1$ из варианционного ряда. 

$\hat{\theta} = x_1$

в нашем случае



In [ ]:
x_1 = sorted([-12.8, -8.85, -13.45, -5.17, -12.11, -16.13, -7.57, -1.64, -8.91, -6.62])[0]
x_1

-16.13

**Ответ:** $\hat{\theta} = -16.13 $

#### Задача №3. Экспоненциальное распределение

По выборке $(x,y,z,t,k,s)$ с помощью метода  максимального правдоподобия найдите точечную оценку параметра $\lambda$ экспоненциально распределенной случайной величины

$ f_\xi(x) =  \begin{cases}
                \lambda e^{-\lambda x} , x \geq 0,
                \\
                0 , x < 0.
              \end{cases} $

**Решение:**

Т.к. экспоненциальное распределение непрерывно, то плотность распределения будет равна $p_\theta = \lambda e^{- \lambda x}$, тогда функция правдоподобия будет равна

$L_X(\lambda) = \prod\limits_{i = 1}^n \lambda e^{- \lambda x_i} = \lambda^n e^{- \lambda \sum\limits_{i=1}^n x_i}$

а логарифмическая функция правдоподобия

$l_X(\lambda) = ln(\lambda^n e^{- \lambda \sum\limits_{i=1}^n x_i}) = n \cdot ln(\lambda) - \lambda \sum\limits_{i=1}^n x_i \rightarrow \underset{\lambda}{\operatorname{max}}$

чтобы найти точку максимума, нужно взять производную

$\frac{\delta l_X}{\delta \lambda} = \frac{n}{\lambda} - \sum\limits_{i=1}^n x_i = 0$

отсюда

$\hat{\lambda} = \frac{n}{\sum\limits_{i=1}^n x_i}$

Проверим, что в данной $\frac{n}{\sum\limits_{i=1}^n x_i}$ точке логарифмическая функция достигает максимума, для этого возьмем вторую частную производную.

$\frac{\delta^2 l_X}{\delta \lambda^2} = n \cdot -\lambda^{-2} = -\frac{n}{\lambda^2}$

подставим $ \frac{n}{\sum\limits_{i=1}^n x_i} $ вместо $\lambda$ 

$-\frac{n}{\frac{n}{\sum\limits_{i=1}^n x_i}} = - \sum\limits_{i=1}^n x_i < 0$

Видим, что вторая производная в точке $\frac{n}{\sum\limits_{i=1}^n x_i}$ меньше нуля, значит логарифмическая функция потерь в данной точке достигает максимума.

**Ответ:**

$\hat{\lambda} = \frac{n}{\sum\limits_{i=1}^n x_i}$

$\hat{\lambda} = \frac{6}{x + y + z + t + k + s}$

#### Задача №4. Биномиальное распределение

Найти методом наибольшего правдоподобия оценку параметра p
биномиального распределения $P_n(k) = C^k_n p^k (1-p)^{n-k}$, если в $n_1$ независимых испытаниях событие $A$ появилось $m_1$ раз и в $n_2$ независимых испытаниях событие $A$ появилось $m_2$ раз.

**Решение:**

Т.к. биномиальное распределение дискретно, то $ p_p = C^k_n p^k (1-p)^{n-k}$. У нас было два независимых исмыпатаний, поэтому функция правдоподобия принимает вид

$L_X(p) = C^{m_1}_{n_1} p^{m_1} (1-p)^{n_1-m_1} \cdot C^{m_2}_{n_2} p^{m_2} (1-p)^{n_2-m_2} = C^{m_1}_{n_1} C^{m_2}_{n_2} p^{m_1 + m_2} (1-p)^{[(n_1+n_2)-(m_1+m_2)]}$

логарифмическая функция правдоподобия

$l_X(p) = ln(C^{m_1}_{n_1} C^{m_2}_{n_2} p^{m_1 + m_2} (1-p)^{[(n_1+n_2)-(m_1+m_2)]}) = \\ 
ln(C^{m_1}_{n_1} C^{m_2}_{n_2}) + (m_1 + m_2) ln(p) + [(n_1+n_2)-(m_1+m_2)] ln(1-p) $

чтобы найти точку максимума, вычислим производную по параметру $p$

$\frac{\delta l_X}{\delta p} = \frac{m_1 + m_2}{p} - \frac{(n_1+n_2)-(m_1+m_2)}{1-p}$

Запишем уравнение правдоподобия, для этого прировняем производную к нулю

$ \frac{m_1 + m_2}{p} - \frac{(n_1+n_2)-(m_1+m_2)}{1-p} = 0 $

решим уравнение относительно параметра $p$, для удобства сделаем замену

$m_1 + m_2 = a$

$n_1 + n_2 = b$

$ \frac{a}{p} - \frac{b - a}{1-p} = 0$

$ \frac{a}{p} = \frac{b - a}{1-p} $

$ \frac{1}{p} = \frac{b - a}{a(1-p)} $

$ \frac{1-p}{p} = \frac{b - a}{a} $

$ \frac{1}{p} - 1 = \frac{b - a}{a} $

$ \frac{1}{p} = \frac{b - a}{a} + 1 $

$ \frac{1}{p} = \frac{b - a + a}{a} $

$ p = \frac{a}{b} $

таким образом, логарифмическая функция правидоподобия достигает максимума в точке

$ p = \frac{m_1 + m_2}{n_1 + n_2} $

возьмем вторую производную, чтобы убедится, что в этой точке $ p = \frac{m_1 + m_2}{n_1 + n_2} $логарифмическая функция потерь достигает максимума

$\frac{\delta^2 l_X}{\delta p^2} =  \frac{\delta}{\delta p} \bigg ( \frac{a}{p} - \frac{b-a}{1-p} \bigg ) = a(-p^{-2}) - (b-a)(\frac{1}{1-p})' = - \frac{a}{p^2} - (b - a) (-1)(\frac{1}{(1-p)^2})(1-p)' = - \frac{a}{p^2} - (b - a) (-1)(\frac{1}{(1-p)^2})(-1) = - \frac{a}{p^2} - \frac{b - a}{(1-p)^2} $

По условию задачи и нашей замене переменных $b>a$ всегда! В числители стоит квадрат, значит числитель всегда положителен, а слагаемые всегда отрицательные. Значит вторая производная в точке $p = \frac{a}{b}$ отричательная, что говорит о том, что это точка максимума.

**Ответ:**

$\hat{p} = \frac{a}{b} = \frac{m_1 + m_2}{n_1 + n_2}$
 


#### Задача №5. Нормальное распределение

Используя метод наибольшего правдоподобия, оценить параметры $а$ и $σ^2$ 
**нормального распределения**, если в результате $n$ независимых испытаний случайная величина $\xi$ приняла значения $\xi_1, ..., \xi_n$.

**Решение:**

Т.к. нормальное распределение непрерывно, значит функция правдоподобия будет иметь вид

$L_X(\theta) = \prod\limits_{i = 1}^n \frac{1}{\sigma \sqrt{2\pi}} \exp{ \bigg ( - \frac{(\xi-a)^2}{2 \sigma^2} \bigg ) } =  \bigg ( \frac{1}{\sigma \sqrt{2\pi}} \bigg )^n \exp \bigg ( - \frac{1}{2} \sum\limits_{i=1}^n \frac{(\xi_i-a)^2}{\sigma^2} \bigg ) = \bigg ( \frac{1}{\sigma \sqrt{2\pi}} \bigg )^n \exp \bigg ( - \frac{1}{2 \sigma^2} \sum\limits_{i=1}^n (\xi_i-a)^2 \bigg )$

логарифмическая функция потерь равна

$l_X =  n \cdot ln(\frac{1}{\sigma \sqrt{2\pi}}) - \frac{1}{2 \sigma^2} \sum\limits_{i=1}^n (\xi_i-a)^2 =  - n \cdot ln(\sigma \sqrt{2\pi}) - \frac{1}{2 \sigma^2} \sum\limits_{i=1}^n (\xi_i-a)^2 = - n \cdot ( ln(\sigma) + ln(\sqrt{2\pi})) - \frac{1}{2 \sigma^2} \sum\limits_{i=1}^n (\xi_i-a)^2 = - n \cdot ln(\sigma) - n \cdot ln(\sqrt{2\pi}) - \frac{1}{2 \sigma^2} \sum\limits_{i=1}^n (\xi_i-a)^2 $

Найдём частные производные:

заметим,
$(\sum\limits_{i=1}^n (\xi_i-a)^2)' = ((\xi_1-a)^2)' + ... + ((\xi_n-a)^2)' = 2(\xi_1-a)(\xi_1-a)' + ... + 2 (\xi_n-a)(\xi_n-a)' = -2(\xi_1-a) - ... - 2 (\xi_n-a) = -2 \sum\limits_{i=1}^n (\xi_i-a)$ 

$ \frac{\delta l_X}{\delta a}  = \frac{1}{\sigma^2} \sum\limits_{i=1}^n (\xi_i-a) $

$ \frac{\delta l_X}{\delta \sigma} = - \frac{n}{\sigma} + \frac{1}{2\sigma^3} \sum\limits_{i=1}^n (\xi_i-a)^2$

Условие экстремума:

$ \frac{1}{\sigma^2} \sum\limits_{i=1}^n (\xi_i-a) = 0 $

$ - \frac{n}{\sigma} + \frac{1}{2\sigma^3} \sum\limits_{i=1}^n (\xi_i-a)^2 = 0 $ заметим, что коэф. $\frac{1}{2}$ во втором слагаемом роли в оптимизации не играет, для упрощения уберем его.

Решим систему:

$\begin{cases}
  \frac{1}{\sigma^2} \sum\limits_{i=1}^n (\xi_i-a) = 0,
  \\
  - \frac{n}{\sigma} + \frac{1}{\sigma^3} \sum\limits_{i=1}^n (\xi_i-a)^2 = 0.
\end{cases}
$

$\begin{cases}
  \sum\limits_{i=1}^n (\xi_i-a) = 0,
  \\
  \frac{1}{\sigma^3} \sum\limits_{i=1}^n (\xi_i-a)^2 = \frac{n}{\sigma}.
\end{cases}
$

$\sum\limits_{i=1}^n (\xi_i-a) = (\xi_1 - a) + ... + (\xi_n - a) = \sum\limits_{i=1}^n \xi_i - an$

$\begin{cases}
  \sum\limits_{i=1}^n \xi_i - an = 0,
  \\
  \frac{1}{\sigma^2} \sum\limits_{i=1}^n (\xi_i-a)^2 = n.
\end{cases}
$

$\begin{cases}
  a = \overline{\xi},
  \\
  \frac{1}{\sigma^2} \sum\limits_{i=1}^n (\xi_i-\overline{\xi})^2 = n.
\end{cases}
$

$\begin{cases}
  a = \overline{\xi},
  \\
  \sigma^2 = \frac{1}{n} \sum\limits_{i=1}^n (\xi_i-\overline{\xi})^2 = D_{выб}.
\end{cases}
$

**Ответ:**

$ \hat{a} = \overline \xi $

$ \hat{\sigma^2} = D_{выб}$

#### Задача №6. Распределение Пуассона

Случайная величина $X$ (число появлений события $A$ в $m$ независимых испытаниях) подчинена закону **распределения Пуассона** с неизвестным параметром $\lambda$ : 

$P_m(X=x_i) = \frac{\lambda^{x_i} e^{-\lambda}}{x_i!}$

где $m$ - число испытаний в одном опыте, $x_i$ - число появлений события в $i$-ом опыте $(i n =1, 2,..., )$.

Найти методом наибольшего правдоподобия по выборке $x_1, ..., x_n$ точечную оценку неизвестного параметра $\lambda$ распределения Пуассона. 


**Решение:**

функция правдоподобия

$ L_X(\lambda) = \prod\limits_{i = 1}^n \frac{\lambda^{x_i} e^{-\lambda}}{x_i!} = \frac{\lambda^{\sum\limits_{i=1}^n x_i} }{\prod\limits_{i = 1}^n x_i!} e^{-\lambda n} $

логарифмическая функция потерь

$ l_X(\lambda) = ln(\lambda^{\sum\limits_{i=1}^n x_i}) - ln(\prod\limits_{i = 1}^n x_i!) + ln(e^{-\lambda n}) = \sum\limits_{i=1}^n x_i ln(\lambda) -  \sum\limits_{i=1}^n ln(x_i!) - \lambda n $

чтобы максимизировать функцию потерь, найдём маскимум, для этого найдем производную

$ \frac{\delta }{\delta \lambda} (\sum\limits_{i=1}^n x_i ln(\lambda) -  \sum\limits_{i=1}^n ln(x_i!) - \lambda n ) = \frac{\sum\limits_{i=1}^n x_i}{\lambda} - n $

уравнение функции правдоподобия

$ \frac{\sum\limits_{i=1}^n x_i}{\lambda} - n = 0 $

выразим искомый параметр $\lambda$

$ \lambda = \overline{x}$

найдем вторую производную, найдём знак в точке \overline{x}, чтобы убедится, что это точка максимума.

$ \frac{\delta }{\delta \lambda} (\frac{\sum\limits_{i=1}^n x_i}{\lambda} - n) = - \frac{\sum\limits_{i=1}^n x_i}{\lambda^2}$

подставив во вторую производную наш параметр $ \lambda = \overline{x} $ увидим, что в точке $ \overline{x} $ вторая производная отрицательная, что говорит о том, что это точка максимума функции.

**Ответ:**

$ \hat{\lambda} = \overline{x}$

#### Задача №7. Экспоненциальное распределение

Случайная величина – время безотказной работы изделия имеет показательное
распределение. В таблице приведены данные по времени работы в часах для 1000 изделий. Найти методом максимального правдоподобия точечную оценку неизвестного
параметра $\lambda$ . 

In [ ]:
T1 = pd.DataFrame({'x_i': [5, 15, 25, 35, 45, 55, 65], 'n_i': [365, 245, 150, 100, 70, 45, 25]}).T
T1

0    1    2    3   4   5   6
x_i    5   15   25   35  45  55  65
n_i  365  245  150  100  70  45  25

**Решение:**

Т.к. экспоненциальное распределение непрерывно, то возьмем плотность распределения $p_\theta = \lambda e^{-\lambda x}$, тогда функция распределения

$ L_X(\lambda) = \prod\limits_{i = 1}^n \lambda e^{-\lambda x} = \lambda^n e^{-\lambda \sum\limits_{i=1}^n x_i}$

логарифмическая функция потерь

$ l_X(\lambda) = n \cdot ln(\lambda) - \lambda  \sum\limits_{i=1}^n x_i $

производная

$ \frac{\delta }{\delta \lambda} (n \cdot ln(\lambda) - \lambda  \sum\limits_{i=1}^n x_i) = \frac{n}{\lambda} -  \sum\limits_{i=1}^n x_i$

уравнение функции потерь

$ \frac{n}{\lambda} -  \sum\limits_{i=1}^n x_i = 0 $

выразим отсюда параметр $\lambda$

$ \hat{\lambda} = \frac{n}{\sum\limits_{i=1}^n x_i}$

Теперь вычислим параметр $\lambda$ для нашего случая

$ \hat{\lambda} = \frac{n}{\sum\limits_{i=1}^n x_i n_i}$

In [ ]:
x_i = T1.iloc[0, :].values
n_i = T1.iloc[1, :].values
n = 1000

In [ ]:
lambda_ = n / sum(x_i * n_i)
lambda_

0.05

**Ответ:**

$ \hat{\lambda} = 0.05 $

#### Задача №8. Геометрическое распределение

Найти методом наибольшего правдоподобия по выборке $x_1, ..., x_n$ точечную
оценку параметра $p$ геометрического распределения:

$P(X=x_i) = (1-p)^{x_i - 1}p $

где $x_i$ - число испытаний, произведенных до появления события, $p$ - вероятность появления события в одном испытании. 

**Решение:**

функция правдоподобия

$L_X(p) = \prod\limits_{i = 1}^n (1-p)^{x_i - 1}p = (1-p)^{\sum\limits_{i=1}^n (x_i - 1)}p^n$

логарифмическая функция

$l_X = \sum\limits_{i=1}^n (x_i - 1) ln(1-p) + n \cdot ln(p)$

производная

$\frac{\delta}{p} (\sum\limits_{i=1}^n (x_i - 1) ln(1-p) + n \cdot ln(p)) = - \frac{\sum\limits_{i=1}^n (x_i - 1)}{(1-p)} + \frac{n}{p}$

уравнение функции правдоподобия

$- \frac{\sum\limits_{i=1}^n (x_i - 1)}{(1-p)} + \frac{n}{p} = 0$

выразим параметр

$ \frac{n}{p} = \frac{\sum\limits_{i=1}^n (x_i - 1)}{(1-p)}$

$ \frac{(1-p)}{p} = \frac{\sum\limits_{i=1}^n (x_i - 1)}{n}$

$ \frac{1}{p} - 1 = \frac{\sum\limits_{i=1}^n (x_i - 1)}{n}$

$ \frac{1}{p} = \frac{\sum\limits_{i=1}^n (x_i - 1)}{n} + 1$

$ \frac{1}{p} = \frac{\sum\limits_{i=1}^n (x_i - 1) + n}{n} $

$ p = \frac{n}{\sum\limits_{i=1}^n (x_i - 1) + n} $

**Ответ:**

$ p = \frac{n}{\sum\limits_{i=1}^n (x_i - 1) + n} $

#### Задача №9. Произвольное распределение

Методом максимального правдоподобия найти точечную оценку параметра λ по
данной выборке 

X 1-3, 3-5, 5-7, 7-9, 9-11, 11-13, 13-15, 15-17, 17-19

n  5    6    7 15 22 27 30 34 35

при условии, что соответствующая непрерывная случайная величина имеет плотность распределения

$ f(x) = \begin{cases}
          \lambda \exp(\lambda(x-20)), x \leq 20
          \\
          0, x > 20.
        \end{cases}$


**Решение:**

Всего провели 181 испытание. Полученные 181 результатов разбили на 9 классов,
исходя из их попадания в один из 9 промежутков $\Delta X$. Обозначим $x^i$ результат, попавший в интервал $\Delta X_i$. Составим функцию правдоподобия:

$ L_X(\lambda) = \prod\limits_{i = 1}^{181} f(x^i) = \prod\limits_{i = 1}^{5} f(x^1) \cdot \prod\limits_{i = 1}^{6} f(x^2) \cdot ... \cdot \prod\limits_{i = 1}^{35} f(x^9) = $ \\

$ \prod\limits_{i = 1}^{5} \lambda e^{(\lambda(x^1_i-20))} \cdot ... \cdot \prod\limits_{i = 1}^{35} \lambda e^{(\lambda(x^9_i-20))} = $ \\

$ \lambda^5 e^{(\lambda \sum\limits_{i=1}^5 (x^1_i-20))} \cdot ... \cdot \lambda^{35} e^{(\lambda \sum\limits_{i=1}^{35} (x^9_i-20))} = $ \\

$ \lambda^5 e^{(\lambda (5x^1-20 \cdot 5))} \cdot ... \cdot \lambda^{35} e^{(\lambda (35x^9-20 \cdot 35))} =  $ \\

$ \lambda^{181} e^{\lambda (5x^1 + 6x^2 + ... + 35x^9 - 20 \cdot 181)} = $ \\

$ \lambda^{181} e^{\lambda (\sum\limits_{i=1}^{181} x_i - 3620)} $

аналогично можно было найти в общем случае

$ L_X(\lambda) = \prod\limits_{i = 1}^n \lambda e^{(\lambda(x_i-20))} = \lambda^n e^{\lambda (\sum\limits_{i=1}^n x_i - 20 \cdot n)}  $

логарифмическая функция потерь

$ l_X = 181 \cdot ln(\lambda)  + \lambda (\sum\limits_{i=1}^{181} x_i - 3620) $

производная

$\frac{\delta}{\delta\lambda} ( 181 \cdot ln(\lambda)  + \lambda (\sum\limits_{i=1}^{181} x_i - 3620) ) = \frac{181}{\lambda} + \sum\limits_{i=1}^{181} x_i - 3620$

уравнение правдоподобия

$ \frac{181}{\lambda} + \sum\limits_{i=1}^{181} x_i - 3620 = 0 $

выразим нужный параметр

$ \lambda = \frac{181}{3620 -  \sum\limits_{i=1}^{181} x_i} = \frac{1}{20 -  \frac{ \sum\limits_{i=1}^{181} x_i}{181} } = \frac{1}{20 - \overline x_в} $

Найдем вторую производную, чтобы проверить что это точка максимума функции

$ \frac{\delta}{\delta\lambda} (\frac{181}{\lambda} + \sum\limits_{i=1}^{181} x_i - 3620) = - \frac{181}{\lambda^{2}} < 0 $

Во вторую производную подставим найденную точку

$ - \frac{181}{(\frac{1}{20 - \overline x_в})^{2}} < 0 $

Т.к. в знаменателе стоит квадрат, то знаменатель всегда будет отрицательный, а вторая производная меньше нуля. Значит в качестве оценки возьмем

$\hat{\lambda} = \frac{1}{20 - \overline x_в} $

Найдём среднее из нашей выборки, возьмем средние значения из интервалов

In [ ]:
T2 = pd.DataFrame({'x_i': [2, 4, 6, 8, 10, 12, 14, 16, 18], 'n_i': [5, 6, 7, 15, 22, 27, 30, 34, 35]}).T
T2

0  1  2   3   4   5   6   7   8
x_i  2  4  6   8  10  12  14  16  18
n_i  5  6  7  15  22  27  30  34  35

$ \overline x_{в} = \frac{\sum\limits_{i=1}^N x_i n_i}{\sum\limits_{i=1}^N n_i} $

In [ ]:
x_i = T2.iloc[0, :]
n_i = T2.iloc[1, :]

In [ ]:
x_mean = round(sum(x_i * n_i) / sum(n_i), 3)
x_mean

12.895

In [ ]:
lambda_ = round(1 / (20 - x_mean), 3)
lambda_

0.141

**Ответ:**

$\hat{\lambda} = \frac{1}{20 - \overline x_в} = 0.141$

#### Задача №10. Функция правдоподобия для пользователя интернета

Рассмотрим следующий эксперимент, имитирующий поведение пользователя  при поиске информации в поисковой системе:

Пользователь последовательно просматривает документы до тех пор, пока не найдет нужную информацию или не устанет.  Вероятность того, что пользователь найдет информацию в документе с номером $i$ равна $p$, вероятность того, что он не найдет информацию в документе $i$ равна $(1-p)$. Вероятность $p$ не зависит от номера документа. Вероятность того, что пользователь устанет после просмотра $i$-го документа равна $g$, соответственно продолжит поиски с вероятностью $(1−g)$. Данная вероятность не зависит от номера просматриваемого документа. Пользователь всегда просматривает как минимум один документ. При этом нам известно, нашел ли пользователь информацию или нет ($F$ - нашел, $N$ - не нашел)

Наблюдение: число просмотренных документов и факт "найдена информация или нет"

Например,

наблюдение $'2 F'$ означает, что пользователь просмотрел первый документ, не нашел в нем информацию, не устал,  просмотрел второй документ и нашел в нем информацию.
наблюдение  $'1 N'$ означает, что пользователь просмотрел первый документ, не нашел в нем информацию и устал, прекратил просматривать документы.

В результате эксперимента мы получили 3 наблюдения:

'2F', '3N', '1F'

Составьте функцию правдоподобия выборки. У вас должна получиться функция $p$ и $g$.

**Решение:**

Из условия задачи понятно, что мы имеем дело с геометрическим распределением, т.к. у нас идёт цепочка событий, которая имеет n-1 неудач, и в конце успех (n - количество попыток). Случайная величина с таким распределением имеет смысл номера первого успешного испытания в схеме Бернулли с вероятностью успеха $p$.

Вероятность геометрического распределения имеет вид:

$P(X=n) = (1-p)^{n-1}p$, где $n$ - это номер попытки в которой будет первый успех.

В нашем случае есть две случайные величины имеющие геометрическое распределение, обозначим их X и Y.

$p$ - нашёл информацию

$1-p$ - не нашёл информацию

$g$ - не устал

$1-g$ - устал

События:

2F - посмотрел 1-ю страницу, не нашёл информацию не устал, посмотрел вторую страницу, нашёл информацию.

3N -  посмотрел 1-ю страницу, не нашёл информация, не устал, посмотрел 2-ю страницу, не нашёл информация, не устал, посмотрел 3-ю страницу не нашёл информацию, устал.

1F - посмотрел первую и нашёл нужную информацию.

$ P(2F) = (1-p)(1-g)p $

$ P(3N) = (1-p)(1-g)(1-p)(1-g)(1-p)g $

$ P(1F) = p $

Т.к. у нас дискретное распределение, то в качестве функции правдоподобия берём вероятность 

$ p_X = P(X=n) = (1-p)^{n-1}p $ 

$ g_Y = P(Y=n) = (1-g)^{n-1}g $ 

Т.к. у нас две случайные величины, то функция правдоподобия будет 

$L_X(p, g) =  p_X \cdot  g_Y = P(2F) \cdot P(3N) \cdot P(1F) = (1-p)^4p^2(1-g)^3q$

**Ответ:**

$L_X(p, g) = (1-p)^4p^2(1-g)^3q$


#### Задача №11. Функция правдоподобия для пользователя интернета (часть 2)

In [ ]:
data = pd.read_csv('data/sample_2_4.txt', header = None).values
data = [ data[i][0] for i in range(len(data)) ]

In [ ]:
sample = np.unique(data, return_counts=True)

In [ ]:
sample

(array(['1 F', '1 N', '2 F', '2 N', '3 F', '3 N', '4 F', '4 N', '5 F',
        '5 N'], dtype='<U3'), array([43, 10, 19,  9, 11,  1,  4,  1,  1,  1]))

In [ ]:
T3 = pd.DataFrame({'x_i': sample[0], 'n_i': sample[1]}).T
T3

0    1    2    3    4    5    6    7    8    9
x_i  1 F  1 N  2 F  2 N  3 F  3 N  4 F  4 N  5 F  5 N
n_i   43   10   19    9   11    1    4    1    1    1

Т.к. гометрическое распределение дискретно, возьмем вероятность для построения функции правдоподобия, предварительно расписав вероятности для каждого значения случайной величины.

$ p_X = P(X=n) = (1-p)^{n-1}p $ 

$ g_Y = P(Y=n) = (1-g)^{n-1}g $

$ P(1F) = p $

$ P(1N) = (1-p)g $

$ P(2F) = (1-p)(1-g)p $

$ P(2N) = (1-p)(1-g)(1-p)g $

$ P(3F) = (1-p)(1-g)(1-p)(1-g)p $

$ P(3N) = (1-p)(1-g)(1-p)(1-g)(1-p)g $

видим закономерность, составим формулу для общего случая

$ P(nF) = (1-g)^{n-1}(1-p)^{n-1}p $

$ P(nN) = (1-p)^n(1-g)^{n-1}g $

применим формулу для последующих последовательностей

$ P(4F) = (1-g)^3(1-p)^3p $

$ P(4N) = (1-p)^4(1-g)^3g $

$ P(5F) = (1-g)^4(1-p)^4p $

$ P(5N) = (1-p)^5(1-g)^4g $


$L_X(p, g) = \prod\limits_{i = 1}^n p_X \cdot \prod\limits_{i = 1}^n g_Y = \prod\limits_{i = 1}^{43} P(1F) \cdot \prod\limits_{i = 1}^{10} P(1N) \cdot ... \cdot P(5N) = $ \\

$ \prod\limits_{i = 1}^{43} p \cdot \prod\limits_{i = 1}^{10} (1-p)g \cdot ... \cdot (1-p)^5(1-g)^4g = $ \\

$ p^{43} \cdot (1-p)^{10}g^{10} \cdot (1-p)^{19}(1-g)^{19}p^{19} \cdot (1-p)^{2*9}(1-g)^9g^9 \cdot  (1-g)^{2*11}(1-p)^{2*11}p^11 \cdot (1-p)^3(1-g)^2g \cdot (1-g)^{3*4}(1-p)^{3*4}p \cdot  (1-p)^4(1-g)^3g \cdot (1-g)^4(1-p)^4p \cdot (1-p)^5(1-g)^4g = $ \\

$ (1-p)^{97} \cdot p^{78} \cdot (1-g)^{75} \cdot g^{22} $

Получили функцию правдоподобия для частного случая

$L_X(p, g) = (1-p)^{97} \cdot p^{78} \cdot (1-g)^{75} \cdot g^{22} $

логарифмическая функция правдоподобия

$l_X(p, g) = 97 \cdot ln((1-p)) + 78 \cdot ln(p) + 75 \cdot ln((1-g)) + 22 \cdot ln(g)$

частные производные логарифмической функции правдоподобия

$ \frac{\delta l_X}{\delta p} =  - \frac{97}{1-p} + \frac{78}{p} $

$ \frac{\delta l_X}{\delta g} =  - \frac{75}{1-g} + \frac{22}{g} $

запишем уравнения правдоподобия для $p$

$ - \frac{97}{1-p} + \frac{78}{p} = 0 $

$ \frac{78}{p} = \frac{97}{1-p} $

$ \frac{1-p}{p} = \frac{97}{78} $

$ \frac{1}{p} - 1 = \frac{97}{78} $

$ \frac{1}{p} = \frac{97}{78} + 1$

$ \frac{1}{p} = \frac{175}{78} $

$ p = \frac{78}{175} \approx 0,448$

запишем уравнения правдоподобия для $g$

$ - \frac{75}{1-g} + \frac{22}{g} = 0 $

$ \frac{22}{g} = \frac{75}{1-g}  $

$ \frac{1-g}{g} = \frac{75}{22}  $

$ \frac{1}{g} = \frac{75}{22} + 1  $

$ \frac{1}{g} = \frac{97}{22} $

$ g = \frac{22}{97} \approx 0,227 $

**Ответ:**

$ \hat{p} \approx 0,446 $

$ \hat{g} \approx 0,227 $


**Общий случай**

Заметим, что оценка зависит от количества функций, поэтому мы можем вывести общую формулу, для дальнейшего подсчёта

m - кол-во (1-p)

k - кол-во p

p - кол-во (1-g)

l - кол-во g

$ \hat{p} = \frac{k}{m + k} $

$ \hat{g} = \frac{l}{l + p} $



In [ ]:
def F(n):
  return {'(1-g)': n-1, 'g': 0, '(1-p)': n-1, 'p': 1}

In [ ]:
def N(n):
  return {'(1-g)': n-1, 'g': 1, '(1-p)': n, 'p': 0}

In [ ]:
def sum_values(dict1, dict2):
  c = Counter(dict1)
  c.update(dict2)
  return c

In [ ]:
def count_function(sample, t, dict):
  coefficient = sample[1][t]
  amount_function = {key: dict[key] * coefficient  for key in dict.keys()}
  return amount_function

In [ ]:
def all_function(path):
  data = pd.read_csv(path, header = None).iloc[:, 0].values
  sample = np.unique(data, return_counts=True)

  all_function = {}
  for t, A in enumerate(sample[0]):
    n = int(A.split()[0])
    letter = A.split()[1]
    if letter == 'F':
      dict_ = F(n)
      amount_function = count_function(sample, t, dict_)
    else:
      dict_ = N(n)
      amount_function = count_function(sample, t, dict_)
    all_function = sum_values(all_function, amount_function)

  return all_function

In [ ]:
def get_evaluations(amount_function):
  m = amount_function['(1-p)']
  k = amount_function['p']

  p_hat = round(k / (k+m), 3)

  p = amount_function['(1-g)']
  l = amount_function['g']

  g_hat = round(l / (l+p), 3)

  return p_hat, g_hat

In [ ]:
path = 'data/sample_2_4.txt'
amount_function = all_function(path)
p_hat, g_hat = get_evaluations(amount_function)
p_hat, g_hat

(0.446, 0.227)

##### Простое решение

$ L = (1 - g)^{\sum\limits_{i \in F \cup N} n_i - 1 } \cdot (1 - p)^{\sum\limits_{i \in F} n_i - 1 +  \sum\limits_{i \in N} n_i } \cdot g^{|N|} \cdot p^{|N|}$

optimal in $ g = \frac{|N|}{|N| + a} $, $p = \frac{|F|}{|F| + b}$

where $ a = \sum\limits_{i \in F \cup N} n_i - 1$, $ b = \sum\limits_{i \in F} n_i - 1 +  \sum\limits_{i \in N} n_i $

In [ ]:
data = []

with open('data/sample_2_4.txt', 'r') as f:
    for line in f:
        n, res = line.split()
        data.append((int(n), res))

a = 0
b = 0
F = 0
N = 0
for n, res in data:

    if res is 'F':
        F += 1
        b += n - 1
    else:
        N += 1
        b += n
    a += n - 1

g = N / (N + a)
p = F / (F + b)

print(f'maximum of L in p = {p:.3f}, g = {g:.3f}')

#### Задача №12. Равномерное распределение

Из генеральной совокупности $\xi$, имеющей следующую плотность распределения

$f_\xi(x) = \begin{cases}
             ax^{(a-1)} \: при \: 0 < x < 1, a > 1,
              \\
             0, иначе
            \end{cases}$

$a$ - неизвестный параметр.

Найдите мат.ожидание случайной величины. У вас должна получиться функция от $a$.


**Решение:**

Т.к. есть плотность вероятности, значит распределение непрерывное. Это равномерное непрерывное распределение. Мат. ожидание непрерывного распределения равно

$ EX = \int\limits_{-\infty}^{+ \infty} f(x) x dx $

В нашем случае

$ E\xi = \int\limits_0^1 ax^{(a-1)} x dx =  \int\limits_0^1 a x^a dx = a \int\limits_0^1 x^a dx = a \frac{x^{(a+1)}}{a+1} |_0^1 = a \frac{1^{(a+1)}}{a+1} - a \frac{0^{(a+1)}}{a+1} = \frac{a}{a+1}$

**Ответ:**

$ E\xi = \frac{a}{a+1}$

#### Задача №13. Равномерное распределение

Из генеральной совокупности $\xi$, имеющей следующую плотность распределения

$f_\xi(x) = \begin{cases}
             ax^{(a-1)} \: при \: 0 < x < 1, a > 1,
              \\
             0, иначе
            \end{cases}$

$a$ - неизвестный параметр.

По выборке 0.4, 0.7, 0.9 найдите функцию правдоподобия $L(X_{[n]},a)$ 

У вас должна получиться функция от $a$.

**Решение:**

Составим функцию правдоподобия

$ L_X(a) = \prod\limits_{i = 1}^{n} ax^{(a-1)}_i = a^n \prod\limits_{i = 1}^{n} x^{a-1}_i $

найдём значение функции правдоподобия от выборки

$ L_X(x, a) = a^n \prod\limits_{i = 1}^{n} x^{a-1}_i = a^3(0.4^{a-1} \cdot 0.7^{a-1} \cdot 0.9^{a-1})$

**Ответ:**

$ L_X(x, a) = a^3(0.4^{a-1} \cdot 0.7^{a-1} \cdot 0.9^{a-1})$

#### Задача №14. Равномерное распределение

Из генеральной совокупности $\xi$, имеющей следующую плотность распределения

$f_\xi(x) = \begin{cases}
             a x^{(a-1)} \: при \: 0 < x < 1, \theta > 1,
              \\
             0, иначе
            \end{cases}$

$a$ - неизвестный параметр.

По выборке 0.4, 0.7, 0.9 найдите оценки параметра $a$, построенные с помощью:

1. метода моментов g(x) = x
2. метода максимального правдоподобия

У вас должна получиться функция от $a$.

**Решение:**

**Метод моментов**

Теоритическое мат. ожидания равномерного распределения равно $EX = \frac{a + b}{2} $ 

В нашем случае

$EX = \frac{0 + 1}{2} = \frac{1}{2}$

но нам нужно найти параметр $a$, поэтому вместо стандартго мат. ожидания воспользуемся найденным в задаче №12

$EX = \frac{a}{a+1}$

тогда

$ EX = \frac{a}{a+1} = \overline x_{выб} =  \frac{1}{n} \sum\limits_{i=1}^n x_i $

выразим параметр $a$

$ \frac{a}{a+1} = \overline x_{выб} $

$ \frac{a+1}{a} = \frac{1}{\overline x_{выб}} $

$ 1 + \frac{1}{a} = \frac{1}{\overline x_{выб}} $

$ \frac{1}{a} = \frac{1}{\overline x_{выб}} - 1$

$ \frac{1}{a} = \frac{1 - \overline x_{выб}}{\overline x_{выб}} $

$ a = \frac{\overline x_{выб} }{1 - \overline x_{выб}} $

найдем по выборке

In [11]:
n = 3
sample = np.array([0.4, 0.7, 0.9])
x_mean = sum(sample) / 3
a_hat = x_mean / (1 - x_mean)
round(a_hat, 2)

2.0

Составим функцию правдоподобия

$ L_X(a) = \prod\limits_{i = 1}^{n} ax^{(a-1)}_i = a^n \prod\limits_{i = 1}^{n} x^{a-1}_i $

логарифмическая функция правдоподобия

$ l_X(a) = n ln(a) + (a-1) \sum\limits_{i=1}^n ln(x_i) $

производная

$ \frac{\delta l_X}{\delta a} = \frac{n}{a} + \sum\limits_{i=1}^n ln(x_i)  $

уравнение правдоподобия

$ \frac{n}{a} + \sum\limits_{i=1}^n ln(x_i) = 0 $

$ \frac{n}{a} = - \sum\limits_{i=1}^n ln(x_i) $

$ a = - \frac{n}{\sum\limits_{i=1}^n ln(x_i)} $

оценка параметра $a$ по выборке


In [8]:
n = 3
sample = np.array([0.4, 0.7, 0.9])
a_hat = - n / sum(np.log(x) for x in sample)
round(a_hat, 2)

2.18

**Ответ:**

ОММ = 2

ОМП = 2.17
